### PR TITLE
Chore/upgrade chai

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
Upgrade chai to v5 and switch to esm output. This PR superceeds #55 

This was a pretty horrible upgrade but it all boilds down to chai (and in the future other dependencies) no longer supporting commonjs modules. The logic of this PR is therefore something like:

1. `chai` no longer supports `commonjs` modules so we need to change the module type by adding `"type": "module"` to `package.json`
2. we also have to change the `tsconfig` to output in native module format so use `ESNext`
3. node when importing native esm modules needs to refer to an exact file path (no directories or names without extensions). Because `tsc` won't mangle import names this necessitates adding `.js` extensions to all local imports (yes typescript understands what this means)
4. When executing tests with `mocha` or using `nodemon` we need to now tell it how to load modules. Usually this involves adding `--loader ts-node/esm` to any node invocations but this now generates a warning in node asking us to switch to the vastly more horrible `--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'`. Reading arround this seems to be where node is heading so 🤦‍♂️ 
5. Because we're using native node `import` rather than `require` at runtime we can no longer load `json` without generating an experimental warning so I've changed that to read the swagger file manually

### CODE REVIEW

**Review by n3op2 — DISMISSED** [review]
hm... as far as I remember this was a bit weird that we need to import from build rather than .ts and specify index file

File: nodemon.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "exec": "node --import 'data:text/javascript,import { register } from \"node:module\"; import { pathToFileURL } from \"node:url\"; register(\"ts-node/esm\", pathToFileURL(\"./\"));' src/index.ts",
```
Comment (dc-andysign) [inline nodemon.json:2]: Oh, this looks horrible ..

File: package.json [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
     "build": "tsoa spec-and-routes && tsc",
     "start": "node build/index.js",
-    "dev": "npm run build && NODE_ENV=dev concurrently \"npx tsc --watch\" \"nodemon -q build/index.js | pino-colada\"",
-    "test:offchain": "NODE_ENV=test ./node_modules/.bin/mocha --config ./test/mocharc.json -r ts-node/register ./test/integration/offchain/*.test.ts",
-    "test:onchain": "NODE_ENV=test ./node_modules/.bin/mocha --config ./test/mocharc.json -r ts-node/register ./test/integration/onchain/*.test.ts",
-    "test:unit": "NODE_ENV=test ./node_modules/.bin/mocha --config ./test/mocharc.json -r ts-node/register ./src/**/*.test.ts",
-    "db:migrate": "npx knex migrate:latest --knexfile src/lib/db/knexfile.ts",
-    "db:rollback": "npx knex migrate:rollback --knexfile src/lib/db/knexfile.ts",
+    "dev": "concurrently \"NODE_ENV=dev nodemon | pino-colada\" \"nodemon -x tsoa spec-and-routes --ignore src/routes.ts --ignore src/swagger.json\"",
+    "test:offchain": "NODE_ENV=test ./node_modules/.bin/mocha --config ./test/mocharc.json ./test/integration/offchain/*.test.ts",
```
Comment (dc-andysign) [inline package.json:14]: What about adding another ENV variable, would that work .. as in 

```
NODE_OPTIONS="--loader ts-node/esm --experimental-json-modules"
```

Comment (mattdean-digicatapult) [reply package.json:14]: These both appear as experimental options and don't seem to be where node is going unfortunately. https://github.com/nodejs/node/issues/51196

File: nodemon.json (hunk 1 above)
Comment (mattdean-digicatapult) [reply nodemon.json:2]: Yep but it's what the node people are suggesting https://github.com/nodejs/node/issues/51196

**Comment by mattdean-digicatapult** [comment]
> hm... as far as I remember this was a bit weird that we need to import from build rather than .ts and specify index file

Yes I need to have another look at the `knex` commands. I think these are the only places I now have issues

**Comment by mattdean-digicatapult** [comment]
https://github.com/knex/knex/issues/4793 does not look good

**Comment by mattdean-digicatapult** [comment]
I've managed to fix this using:

```json
{
  "db:cmd": "node --import='data:text/javascript,import { register } from \"node:module\"; import { pathToFileURL } from \"node:url\"; register(\"ts-node/esm\", pathToFileURL(\"./\"));' ./node_modules/.bin/knex",
  "db:migrate": "npm run db:cmd -- migrate:latest --knexfile src/lib/db/knexfile.ts",
  "db:rollback": "npm run db:cmd -- migrate:rollback --knexfile src/lib/db/knexfile.ts"
}
```

**Comment by mattdean-digicatapult** [comment]
Broadly speaking I dislike all the horribleness especially needing to wrap all node commands in essentially:

```
--import='data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'
```

This is basically caused by ts-node being massively behind the curve https://github.com/TypeStrong/ts-node/issues/1997. As they don't seem to be responding it might be worth looking for a ts-node alternative and the two best options I've seen are:

* tsx https://github.com/privatenumber/tsx
* tsimp https://github.com/tapjs/tsimp

I'm not sure I want to do this switch right now, but we can consider it

**Review by dc-andysign — APPROVED** [review]
Looks good to me 👍

**Review by n3op2 — APPROVED** [review]

**Comment by mattdean-digicatapult** [comment]
closing in favour of #66

**Pull request closed without merging**